### ISSUE TITLE
kdu_expand hanging

### ISSUE THREAD

**Issue by bcail** [loris-imageserver/loris#253]
Sometimes when I run the loris test suite (on Linux Mint 17.3), it hangs - I think it's when it's trying to run the kdu_expand command. If I kill the test process and run the tests again, they work.

Also, in production we're seeing that sometimes a kdu_expand process will just hang (until apache is restarted).

Is anyone else seeing these issues?

**Comment by jpstroop** [comment]
It's a tough issue; I wrote this up on SO last week: http://stackoverflow.com/q/40352825/714478

**Comment by bcail** [comment]
Hmm. One commenter mentioned using a temporary file instead of a named pip - are there issues with that approach?

**Comment by jpstroop** [comment]
The only issue is that it's a lot of disk i/o which I'm sure would slow things down.

**Comment by jpstroop** [comment]
I wonder if it could be a config option. Seems like it's a little deep under the hood to ask average users to understand, but other than adding some preflight checks and hoping for the best, I've never really been sure how to handle this.

**Comment by jpstroop** [comment]
Reading your issue more closely, I guess I haven't seen it hang randomly--it's always when `kdu_expand` is misconfigured or missing, or the jp2 is corrupt.

**Comment by bcail** [comment]
Looks like this happened on the Travis tests for PR #256. The tests froze for 10 minutes and then were killed. Restarting the tests made them run successfully to completion.

**Comment by bcail** [comment]
@jpstroop I've been looking at this issue (it happens regularly in our production environment). Does it happen in your system as well?

 I'm attaching a patch where I changed transform() to create a new subprocess, and then kill that subprocess if it times out (hangs). Does that idea seem like something that would work? Would there be issues with thread-safety if loris is running in a threaded environment, or would this be ok because it's sending the terminate signal to a whole new process? Any thoughts on this?

[multiprocessing.txt](https://github.com/loris-imageserver/loris/files/745204/multiprocessing.txt)

**Comment by jpstroop** [comment]
Hey. I'm pretty tied up today, but will try to get a look tonight or tomorrow. I have to say, I don't have any evidence that this happens in our environment, but this risk has always bothered me, and I'd be happy to have a solution.

**Comment by jpstroop** [comment]
Do you want to push up a PR? We can always close it if it doesn't work out?

**Comment by bcail** [comment]
Sure, will do. I'll run it in our production environment and make sure the hung processes go away, too.

**Comment by scossu** [comment]
Hi there, 

I'm not sure how much this would add to the conversation but I have encountered this recently too. 

When I send a request for a transform that is not cached, it hangs indefinitely. The Loris log shows in its last line: 

    /opt/kdu/kdu/kdu_expand -quiet -i "/usr/local/loris/cache/1b/66b/164/af6/0aa/beb/6f5/7ff/e20/2e4/1d4/loris_cache.jp2" -num_threads 4  -reduce 4 -o /usr/local/loris/tmp/jp2/zljyn.bmp

If I run the same command on the shell, it succeeds (if I haven't killed the connection I also see the feedback in the loris log) and the image is returned to the original request. 

My environment is not quite standard: 

- RHEL 6 
- Python 2.7.8
- Kakadu libs and binaries installed in /opt/kdu/kdu
- glibc 2.14.1 in /opt (RHEL6 provides 2.12)
- LD_LIBRARY_PATH to find glibc and kdu is set in the cgi script (in order to replicate the kdu_expand command on the shell, I set the env manually)

I would not be surprised if something went wrong on the system or app setup level, which would confirm @jpstroop 's point, but in any case I would find that the application should raise an exception instead of hanging.
